src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/SystemModulesPlugin.java

```diff
@@ -680,25 +692,138 @@ private void genIncubatorModules(ClassBuilder clb) {
-                        }
-                        cob.aload(MD_VAR)
+                        cob.new_(arrayListClassDesc)
+                           .dup()
+                           .constantInstruction(moduleInfos.size())
+                           .invokespecial(arrayListClassDesc, INIT_NAME, MethodTypeDesc.of(CD_void, CD_int))
+                           .astore(DEDUP_LIST_VAR);
+                        cob.aload(0)
+                           .aload(MD_VAR)
+                           .aload(DEDUP_LIST_VAR)
+                           .invokevirtual(
+                                   this.classDesc,
+                                   helperMethodNamePrefix + "0",
+                                   MethodTypeDesc.of(CD_void, CD_MODULE_DESCRIPTOR.arrayType(), arrayListClassDesc)
+                           )
                            .areturn();
                     });
+
+            for (int n = 0, count = 0; n < splitModuleInfos.size(); count += splitModuleInfos.get(n).size(), n++) {
+                int index = n;       // the index of which ModuleInfo being processed in the current batch
+                int start = count;   // the start index to the return ModuleDescriptor array for the current batch
+                clb.withMethodBody(
+                        helperMethodNamePrefix + index,
+                        MethodTypeDesc.of(CD_void, CD_MODULE_DESCRIPTOR.arrayType(), arrayListClassDesc),
+                        ACC_PUBLIC,
+                        cob -> {
+                            cob.aload(2)
+                               .astore(DEDUP_LIST_VAR);
+
+                            if (nextLocalVar > firstVar) {
+                                for (int i = firstVar; i < nextLocalVar; i++) {
+                                    cob.aload(DEDUP_LIST_VAR)
+                                       .constantInstruction(i - firstVar)
+                                       .invokevirtual(arrayListClassDesc, "get", MethodTypeDesc.of(CD_Object, CD_int))
+                                       .astore(i);
+                                }
+                            }
+
+                            List<ModuleInfo> currentBatch = splitModuleInfos.get(index);
+                            for (int j = 0; j < currentBatch.size(); j++) {
+                                ModuleInfo minfo = currentBatch.get(j);
+                                new ModuleDescriptorBuilder(cob,
+                                                            minfo.descriptor(),
+                                                            minfo.packages(),
+                                                            start + j).build();
+                            }
+
+                            if (index < splitModuleInfos.size() - 1) {
+                                if ((nextLocalVar > firstVar) && (nextLocalVar > wrapper.lastCopiedVar)) {
+                                    for (int i = wrapper.lastCopiedVar + 1; i < nextLocalVar; i++) {
```

Comment:
shouldn't it start with `wrapper.lastCopiedVar`? 

```suggestion
                            if (index < splitModuleInfos.size() - 1) {
                                if (nextLocalVar > firstLocalVar && nextLocalVar > curLocalVar) {
                                    for (int i = curLocalVar; i < nextLocalVar; i++) {
```